***Project: Exploring eBay Car Sales Data***


In this project, we'll be working with a dataset of used cars from eBay, a classifieds section of the german eBay webiste. 

The aim of this project is to clean the dataset and perform some inital analysis on it. 

In [82]:
#importing pandas and NumPy

import pandas as pd 
import numpy as np

In [83]:
#reading the auto.csv file 

autos = pd.read_csv("autos.csv", encoding = "latin1")

In [84]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [85]:
#display information about to autos dataframe as well as the first few rows
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

**Data overview:**

The following was observed:

   1. There are 15 columns with the datatype described as object, which indicates data stored as strings. 
   2. There are 5 columns that store integer data. 

**Cleaning Column Names:**

We will convert the column names from camelcase to snakecase and change some of the column names based on the data dictionary in order to be more descriptive

In [86]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [87]:
autos.rename(inplace = True, columns={
    'dateCrawled' : 'date_crawled' , 
    'offerType' : 'offer_type' , 
    'vehicleType' : 'vehicle_type' , 
    'yearOfRegistration' : 'registration_year' , 
    'powerPS' : 'power_ps' , 
    'monthOfRegistration' : 'registration_month' , 
    'fuelType' : 'fuel_type' , 
    'notRepairedDamage' : 'unrepaired_damage' , 
    'dateCreated' : 'date_created' , 
    'nrOfPictures' : 'pictures_nr' , 
    'postalCode' : 'postal_code' , 
    'lastSeen' : 'last_seen' , })

In [88]:
#assigning the modified column back to DataFrame.columns attribute

autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'date_created', 'pictures_nr', 'postal_code',
       'last_seen'],
      dtype='object')

In [89]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage         date_created  pictures_nr  \
0     peugeot              nein  2016-03-26 00:00:00            0   
1         bmw              nein  2016-04-04 00:00:00            0   
2  volkswagen              nein  2016-03-26 00:00:00            0   
3       smart              nein  2016-03-12 00:00:00            0   
4        ford              nein  2016-04-01 00:00:00            0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

**Column Title Adjustments**

To enhance readability, we made changes to the column titles of the autos DataFrame. Specifically, we converted the titles from camelCase to snake_case. Additionally, we shortened some of the lengthy column names to improve readability and understanding of the data.

These adjustments help ensure that the column names are consistent, easy to read, and more convenient to work with during the data analysis process.


In [90]:
autos.describe(include = 'all')


date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-04-02 11:37:04  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

               date_created  pictures_nr   postal_code            last_seen  
count                 50000      50000.0  50000.000000                50000  
unique                   76          NaN           NaN                39481  
top     2016-04-03 00:00:00          NaN           NaN  2016-04-07 06:17:27  
freq                   1946          NaN           NaN                    8  
mean                    NaN          0.0  50813.627300                  NaN  
std                     NaN          0.0  25779.747957                  NaN  
min                     NaN          0.0   1067.000000                  NaN  
25%                     NaN          0.0  30451.000000                  NaN  
50%                     NaN          0.0  49577.000000                  NaN  
75%                     NaN          0.0  71540.000000                  NaN  
max                     NaN          0.0  99998.000000                  NaN

**Inital Exploration and Cleaning**
We see some potentially erroneous data the following columns that warrant further investigation:

registration_year - has a minimum and maximum value of 1000 and 9999, respectively.

power_ps - has a maximum value of 17700

registration_month - has a minimum value of 0

We also noted some text columns that need to be converted to numeric:

price
odometer

The num_of_pictures column seems to only contain the value 0, so we can drop this column.
Both the seller and offer_type columns contain the same value in almost all of the rows. These can be dropped as well. 



In [91]:
#Removing $ and , from price column
autos["price"] = autos["price"].str.replace("$" , "").str.replace(",", "").astype(float)
print(autos["price"].head(7))



0    5000.0
1    8500.0
2    8990.0
3    4350.0
4    1350.0
5    7900.0
6     300.0
Name: price, dtype: float64


/tmp/ipykernel_64/1878696660.py:2: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  autos["price"] = autos["price"].str.replace("$" , "").str.replace(",", "").astype(float)


In [92]:
#Removing km and , from the odometer column
autos["odometer"] = autos["odometer"].str.replace(",", "").str.replace("km", "").replace(" ", "")
autos["odometer"] = autos["odometer"].str.rstrip('km').astype(int)
                                                        


autos["odometer"] = autos["odometer"]
print(autos["odometer"].head())


0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer, dtype: int64


In [93]:
autos = autos.rename(columns = {"odometer" : "odometer_km"})

In [94]:
#Exploring the price column for possible outliers

print("unique values:", autos['price'].unique().shape[0]) #shows unique values
print( autos['price'].describe()) #to view min/mex/median/mean
print(autos['price'].value_counts()) #Each price value and their counts


# In order to find the amount of highest price and lowest price and their counts,
#sorting can be done using series.sort_index() method.
print(autos['price'].value_counts().sort_index(ascending = True).head(20))
print(autos['price'].value_counts().sort_index(ascending = False).head(20))

unique values: 2357
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64
0.0        1421
500.0       781
1500.0      734
2500.0      643
1000.0      639
           ... 
414.0         1
79933.0       1
5198.0        1
18890.0       1
16995.0       1
Name: price, Length: 2357, dtype: int64
0.0     1421
1.0      156
2.0        3
3.0        1
5.0        2
8.0        1
9.0        1
10.0       7
11.0       2
12.0       3
13.0       2
14.0       1
15.0       2
17.0       3
18.0       1
20.0       4
25.0       5
29.0       1
30.0       7
35.0       1
Name: price, dtype: int64
99999999.0    1
27322222.0    1
12345678.0    3
11111111.0    2
10000000.0    1
3890000.0     1
1300000.0     1
1234566.0     1
999999.0      2
999990.0      1
350000.0      1
345000.0      1
299000.0      1
295000.0      1
265000.0      1
259000.0      1
250000.0      1
22

Based on the above results there are 2357 categories of prices,

It can be determined the selling prices of the cars started with a minimum of 0.00 and rises up to the maximum value of 100.00 Million.

It also has 1427 counts of 0.00 prices,which is impossible to happen as it does not contain a monetary value for the selling price.

It can be noticed that after 350,000.00, the price values rises up to totally unrealistic values like $12345678.00 where there were three cars sold in that price.
Considering above observations, 0.00 price rows should be dropped from the dataset and only rows in the price range of 1.00 to 350,000 should be kept for the analysis.

In [95]:
autos = autos[autos["price"].between(1, 350000)]
autos['price'].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

In [96]:
#Exploring the odometer_km column for possible outliers
print('Unqiue values:', autos['odometer_km'].unique().shape[0])
print(autos['odometer_km'].describe())
print(autos['odometer_km'].value_counts().sort_index(ascending = True).head(13))


Unqiue values: 13
count     48565.000000
mean     125770.101925
std       39788.636804
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64
5000        836
10000       253
20000       762
30000       780
40000       815
50000      1012
60000      1155
70000      1217
80000      1415
90000      1734
100000     2115
125000     5057
150000    31414
Name: odometer_km, dtype: int64


Based on above results there are only 13 categories of the milleage:

It can be noticed that the pattern of the milleage and count of vehichles with that milleage has risen up gradually according to a consistent pattern.
So there are no outliers that should be removed.

**Exploring the Date Columns**

In [97]:
autos[['date_crawled', 'date_created', 'last_seen']][0:5]

date_crawled         date_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [98]:
autos["date_crawled"].str[:10].value_counts(normalize=True,dropna=False).sort_index()

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

**date_crawled observations:**

From when the ad was first crawled, the traffic was roughly the same from 2016-03-05 to 2016-04-05. After it dropped signifigantly.

In [99]:
autos["last_seen"].str[:10].value_counts(normalize=True,dropna=False).sort_index()

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

In [100]:
autos["date_created"].str[:10].value_counts(normalize=True,dropna=False).sort_index()

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038855
2016-04-04    0.036858
2016-04-05    0.011819
2016-04-06    0.003253
2016-04-07    0.001256
Name: date_created, Length: 76, dtype: float64

In [101]:
autos['registration_year'].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

**Registraton_Year Observations:**

The minimum registration year is 1000.0, which is before cars were invented.

The max is 9999.0, which is in the distant future. 

In [102]:
#Determining the highest and lowest acceptable registration year values
print(autos['registration_year'].value_counts().sort_index(ascending=True).head(27))
print(autos['registration_year'].value_counts().sort_index(ascending=False).head(17))

1000     1
1001     1
1111     1
1800     2
1910     5
1927     1
1929     1
1931     1
1934     2
1937     4
1938     1
1939     1
1941     2
1943     1
1948     1
1950     3
1951     2
1952     1
1953     1
1954     2
1955     2
1956     4
1957     2
1958     4
1959     6
1960    23
1961     6
Name: registration_year, dtype: int64
9999       3
9000       1
8888       1
6200       1
5911       1
5000       4
4800       1
4500       1
4100       1
2800       1
2019       2
2018     470
2017    1392
2016    1220
2015     392
2014     663
2013     803
Name: registration_year, dtype: int64


Based on the registration_year data, acceptable data range is between 1910 and 2016. 

This is based on 1800 and 2800 being un-acceptable registration years.

In [103]:
autos = autos[autos["registration_year"].between(1910,2016)]
autos['registration_year'].value_counts(normalize=True)
print(autos['registration_year'].describe())

count    46681.000000
mean      2002.910756
std          7.185103
min       1910.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64


**Adjusted registration_year observations:**

Once the values outside the upper and lower bounds were removed, it was concluded that most cars were registered in 2002. 

**Exploring Price by Brand**

In [104]:
print(autos['brand'].value_counts().head(20))
print(autos['brand'].unique())
print(autos['brand'].index)


volkswagen        9862
bmw               5137
opel              5022
mercedes_benz     4503
audi              4041
ford              3263
renault           2201
peugeot           1393
fiat              1197
seat               853
skoda              766
nissan             713
mazda              709
smart              661
citroen            654
toyota             593
hyundai            468
sonstige_autos     458
volvo              427
mini               409
Name: brand, dtype: int64
['peugeot' 'bmw' 'volkswagen' 'smart' 'ford' 'chrysler' 'seat' 'renault'
 'mercedes_benz' 'audi' 'sonstige_autos' 'opel' 'mazda' 'porsche' 'mini'
 'toyota' 'dacia' 'nissan' 'jeep' 'saab' 'volvo' 'mitsubishi' 'jaguar'
 'fiat' 'skoda' 'subaru' 'kia' 'citroen' 'chevrolet' 'hyundai' 'honda'
 'daewoo' 'suzuki' 'trabant' 'land_rover' 'alfa_romeo' 'lada' 'rover'
 'daihatsu' 'lancia']
Int64Index([    0,     1,     2,     3,     4,     5,     6,     7,     8,
                9,
            ...
            49990, 49991

**The car brands I am choosing to focus on is:**
- volkswagen
- bmw
- opel
- mercedes benz
- audi

These brands were chosen due to being in the top 8%.

In [105]:
brands = autos["brand"].unique()
brand_mean_price = {}

for b in brands:
    mean_price = autos[autos["brand"]==b]["price"].mean()
    brand_mean_price[b] = mean_price


# Converting the dictionary into a list 
# and sorting it alphabetically by brands.
brand_mean_price_list = sorted(brand_mean_price.items())

# swapping the list
brand_mean_price_list_swapped = [(y, x) for x, y in brand_mean_price_list]

# sorting it based on the values
brand_mean_price_list_swapped_sorted = sorted(brand_mean_price_list_swapped)

# undo the swapping
brand_mean_price_list_sorted = [(y, x) for x, y in brand_mean_price_list_swapped_sorted]

# converting it to a dictionary again in case it is needed
mean_price_brand = {}
for key, value in brand_mean_price_list_sorted:
    mean_price_brand[key] = value

brand_mean_price_list_swapped_sorted

[(1049.0, 'daewoo'),
 (1602.2903225806451, 'rover'),
 (1636.1965811965813, 'daihatsu'),
 (1790.8615384615384, 'trabant'),
 (2474.8646069968195, 'renault'),
 (2688.296296296296, 'lada'),
 (2813.748538011696, 'fiat'),
 (2975.2419354838707, 'opel'),
 (3094.0172290021537, 'peugeot'),
 (3211.6493506493507, 'saab'),
 (3376.22, 'lancia'),
 (3394.5729166666665, 'mitsubishi'),
 (3465.743902439024, 'chrysler'),
 (3580.2239031770046, 'smart'),
 (3749.4695065890287, 'ford'),
 (3779.1391437308866, 'citroen'),
 (3953.29, 'subaru'),
 (4087.690322580645, 'alfa_romeo'),
 (4096.732851985559, 'suzuki'),
 (4107.857923497268, 'honda'),
 (4112.596614950635, 'mazda'),
 (4397.230949589683, 'seat'),
 (4743.40252454418, 'nissan'),
 (4946.501170960188, 'volvo'),
 (5167.091062394604, 'toyota'),
 (5365.254273504273, 'hyundai'),
 (5402.410261610221, 'volkswagen'),
 (5915.528455284553, 'dacia'),
 (5982.330303030303, 'kia'),
 (6368.0, 'skoda'),
 (6684.139097744361, 'chevrolet'),
 (8332.820517811953, 'bmw'),
 (8628.45

**Takeaways**

Based on the top 8% brands of cars; Audi, BMW & Mercedes are the most expensive. Opel are less expensive and Volkswagen is inbetween.


**Sorting Aggregate Data in a DataFrame**

In [106]:
brands = autos["brand"].unique()
brand_mean_miles = {}

for b in brands:
    mean_miles = autos[autos["brand"]==b]["odometer_km"].mean()
    brand_mean_miles[b] = mean_miles
brand_mean_miles

{'peugeot': 127153.62526920316,
 'bmw': 132572.51313996495,
 'volkswagen': 128707.15879132022,
 'smart': 99326.77760968229,
 'ford': 124266.01287159056,
 'chrysler': 132378.0487804878,
 'seat': 121131.30128956624,
 'renault': 128071.33121308497,
 'mercedes_benz': 130788.36331334666,
 'audi': 129157.38678544914,
 'sonstige_autos': 89956.33187772926,
 'opel': 129310.0358422939,
 'mazda': 124464.03385049365,
 'porsche': 96853.14685314686,
 'mini': 88105.13447432763,
 'toyota': 115944.35075885328,
 'dacia': 84268.29268292683,
 'nissan': 118330.99579242637,
 'jeep': 127122.64150943396,
 'saab': 144415.58441558442,
 'volvo': 138067.9156908665,
 'mitsubishi': 126575.52083333333,
 'jaguar': 124178.08219178082,
 'fiat': 117121.9715956558,
 'skoda': 110848.5639686684,
 'subaru': 125300.0,
 'kia': 112530.30303030302,
 'citroen': 119694.18960244648,
 'chevrolet': 99191.72932330827,
 'hyundai': 106442.30769230769,
 'honda': 122172.13114754099,
 'daewoo': 121642.85714285714,
 'suzuki': 108465.703971

In [115]:
#Converting dictionaries to series object, using the series constructor
prices_series = pd.Series(brand_mean_prices)

In [114]:
prices_df = pd.DataFrame(prices_series, columns=['mean_price'])
prices_df

mean_price
peugeot               3094
bmw                   8332
volkswagen            5402
smart                 3580
ford                  3749
chrysler              3465
seat                  4397
renault               2474
mercedes_benz         8628
audi                  9336
sonstige_autos       12338
opel                  2975
mazda                 4112
porsche              45643
mini                 10613
toyota                5167
dacia                 5915
nissan                4743
jeep                 11650
saab                  3211
volvo                 4946
mitsubishi            3394
jaguar               11635
fiat                  2813
skoda                 6368
subaru                3953
kia                   5982
citroen               3779
chevrolet             6684
hyundai               5365
honda                 4107
daewoo                1049
suzuki                4096
trabant               1790
land_rover           19108
alfa_romeo            4087
lada                  2688
rover                 1602
daihatsu              1636
lancia                3376

In [122]:
miles_series = pd.Series(brand_mean_miles)
miles_df = pd.DataFrame(miles_series, columns=['mean_miles'])

mean_mileage_price_brand_df = pd.concat([prices_df, miles_df], axis=1)
mean_mileage_price_brand_df.columns = ["mean_price", "mean_mileage"]
selected = mean_mileage_price_brand_df.loc[["volkswagen","audi","opel","mercedes_benz","bmw"]]
selected

mean_price   mean_mileage
volkswagen           5402  128707.158791
audi                 9336  129157.386785
opel                 2975  129310.035842
mercedes_benz        8628  130788.363313
bmw                  8332  132572.513140

**Conclusion**

As observed, all our focused brands exhibit similar average mileage. Opel stands out as the most affordable option for buyers, appealing to budget-conscious shoppers. On the other end of the spectrum, Audi, Mercedes Benz, and BMW fall into the category of luxury vehicles. Notably, Volkswagen represents a significant portion in our dataset. While not the priciest, Volkswagen strikes a balance between affordability and quality.